### DISCUSSION TITLE
Building without MKL

### DISCUSSION THREAD

**Discussion by kkly1995** [mir-group/pair_nequip#22] [Q&A]
I am trying to build LAMMPS with pair_nequip on a machine with IBM POWER9 CPUs, which to my knowledge is incompatible with MKL. On the other hand, PyTorch and LAMMPS work just fine on these machines (I have NequIP working there). Naturally, when attempting to build, I get the following error:

```
CMake Error in CMakeLists.txt:
  Imported target "torch" includes non-existent path

    "MKL_INCLUDE_DIR-NOTFOUND"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in CMakeLists.txt:
  Target "lammps" contains relative path in its
  INTERFACE_INCLUDE_DIRECTORIES:

    "MKL_INCLUDE_DIR-NOTFOUND"
```

Is there a way to circumvent this? Is MKL strictly necessary? I can see that it is "torch" that is requesting MKL here, but it shouldn't need it, right?

**Comment by Linux-cpp-lisp** [comment]
Hi @kkly1995 ,

This is a good question but one that is for PyTorch, not us-- you want to look up how to build PyTorch / `libtorch` on POWER9, which should presumably involve the appropriate `cmake` options to disable MKL. To the best of my knowledge there is no pre-built POWER build of PyTorch, so you might want to ask whoever built PyTorch on your system in the first place how they did it.

See ex.
https://discuss.pytorch.org/t/can-we-build-libtorch-without-mkl/44612/2
https://github.com/pytorch/pytorch/issues/69068
https://gauravm.gitbook.io/about/blogs/installing-pytorch-and-transformers-on-ibm-powerpc-architecture

If you have any success and it occurs to you to post, it would be great to know what works for others who encounter this issue.

Thanks!

**Reply by Linux-cpp-lisp** [reply]
https://github.com/pytorch/pytorch/issues/69068

is probably the most relevant if you are using `libtorch` from your PyTorch installation and not from a separate new build/download

**Reply by kkly1995** [reply]
I just asked the admin how he built PyTorch and he said he just used a pre-built from Open-CE. I tried building pair_nequip with this particular environment (since it's the same one I have NequIP working on) but it seems to struggle with protobuf:

```
CMake Error at /opt/miniconda3/envs/opence-v1.5.1/lib/python3.9/site-packages/torch/share/cmake/Caffe2/Caffe2Config.cmake:58 (message):
  Your installed Caffe2 version uses protobuf but the protobuf library cannot
  be found.  Did you accidentally remove it, or have you set the right
  CMAKE_PREFIX_PATH? If you do not have protobuf, you will need to install
  protobuf and set the library path accordingly.
Call Stack (most recent call first):
  /opt/miniconda3/envs/opence-v1.5.1/lib/python3.9/site-packages/torch/share/cmake/Torch/TorchConfig.cmake:68 (find_package)
  CMakeLists.txt:922 (find_package)
```

**Reply by Linux-cpp-lisp** [reply]
That sounds more promising (you can always try to `conda install` protobuf), but I've never seen any of this or tried to build on POWER9 before... sorry that I don't have more definitive answers.

I would also take a look at that PyTorch issue I posted, there seemed to be some kind of hack to make it build without MKL. If it doesn't work, I would also reach out to the PyTorch team either by GitHub, their PyTorch Discussions forum, or the PyTorch Slack.
